modules/images/dominant-color/class-dominant-color-image-editor-gd.php

```diff
@@ -0,0 +1,77 @@
+<?php
+/**
+ * WordPress Image Editor Class for Image Manipulation through GD
+ * with dominant color detection
+ *
+ * @package performance-lab
+ * @group dominant-color
+ *
+ * @since n.e.x.t
+ */
+
+/**
+ * WordPress Image Editor Class for Image Manipulation through GD
+ * with dominant color detection
+ *
+ * @since n.e.x.t
+ *
+ * @see WP_Image_Editor
+ */
+class Dominant_Color_Image_Editor_GD extends WP_Image_Editor_GD {
+
+	/**
+	 * Get dominant color from a file.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $default_color Optional. Hex color string, without leading #. Default is light grey.
+	 * @return string Dominant hex color string.
+	 */
+	public function get_dominant_color( $default_color = 'eee' ) {
+
+		if ( ! $this->image ) {
+
+			return $default_color;
+		}
+
+		$shorted_image = imagecreatetruecolor( 1, 1 );
+		imagecopyresampled( $shorted_image, $this->image, 0, 0, 0, 0, 1, 1, imagesx( $this->image ), imagesy( $this->image ) );
+
+		$hex = dechex( imagecolorat( $shorted_image, 0, 0 ) );
+
+		return ( '0' === $hex ) ? $default_color : $hex;
+
+	}
+
+
+	/**
+	 * Looks for transparent pixels in the image.
+	 * If there are none, it returns false.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool
+	 */
+	public function get_has_transparency() {
+
+		if ( ! $this->image ) {
+
+			return false;
+		}
+
+		// walk through the pixels.
+		$w = imagesx( $this->image );
+		$h = imagesy( $this->image );
```

Comment:
I keep looking at this.
We could use the thumbnail but that is a cropped image so we may miss a transparent pixel in the search 
Look for the best solution :-)

Reply:
When uploading images on this branch I noticed I was getting additional PHP timeouts, uploads would succeed only after multiple retries.

Testing this locally I added timing logging to this function and discovered for larger image it alone takes close to 20 seconds (on my well-resourced machine). Based on this, I feel like this may break uploading or greatly reduce upload speed for most users.

Can this use the same approach as the `Imagick` implementation below? Eg, resize the image to one pixel, then use that pixel color?

Reply:
note: if uploading succeeds, then PHP times out during this calculation, media upload will wind up in a failed state.